vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go

```diff
@@ -0,0 +1,149 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllerfetcher
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	autoscalingapi "k8s.io/api/autoscaling/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var now = time.Now
+
+type scaleCacheKey struct {
+	namespace     string
+	groupResource schema.GroupResource
+	name          string
+}
+
+type scaleCacheEntry struct {
+	refreshAfter time.Time
+	deleteAfter  time.Time
+	resource     *autoscalingapi.Scale
+	err          error
+}
+
+// Cache for responses to get queries on controllers. Thread safe.
+// Usage:
+// - `Get` cached response. If there is one use it, otherwise make query and
+// - `Insert` the response you got into the cache.
+// Start two go routines:
+// - One for refereshing cache entries, which calls `GetKeysToRefresh` then for
+//   each key makes query to the API server and calls `Refresh` to update
+//  content of the cache.
+// - Second from removing stale entries which periodically calls `RemoveExpired`
+// Each entry is refreshed after duration (`betweenRefreshes` + `jitter`) passes
+// and is removed if there are no reads for it for more than `lifeTime`.
+type controllerCacheStorage struct {
+	cache            map[scaleCacheKey]scaleCacheEntry
+	mux              sync.Mutex
+	betweenRefreshes time.Duration
+	refreshJitter    time.Duration
+	lifeTime         time.Duration
+}
+
+// Returns bool indicating whether the entry was present in the cache and the cached response.
+// Updates deleteAfter for the element.
+func (cc *controllerCacheStorage) Get(namespace string, groupResource schema.GroupResource, name string) (ok bool, controller *autoscalingapi.Scale, err error) {
+	key := scaleCacheKey{namespace: namespace, groupResource: groupResource, name: name}
+	cc.mux.Lock()
+	defer cc.mux.Unlock()
+	r, ok := cc.cache[key]
+	if ok {
+		r.deleteAfter = now().Add(cc.lifeTime)
+		cc.cache[key] = r
+	}
+	return ok, r.resource, r.err
+}
+
+// If key is in the cache it updates the cached value, error and refresh time (but not time to remove).
+// If the key is missing from the cache does nothing (relevant when we asynchronously updating cache and
+// removing stale entries from it, to avoid adding back an entry which we just removed).
+func (cc *controllerCacheStorage) Refresh(namespace string, groupResource schema.GroupResource, name string, controller *autoscalingapi.Scale, err error) {
+	key := scaleCacheKey{namespace: namespace, groupResource: groupResource, name: name}
+	cc.mux.Lock()
+	defer cc.mux.Unlock()
+	old, ok := cc.cache[key]
+	if !ok {
+		return
+	}
+	jitter := time.Duration(rand.Float64()*float64(cc.refreshJitter.Nanoseconds())) * time.Nanosecond
+	now := now()
+	cc.cache[key] = scaleCacheEntry{
+		refreshAfter: now.Add(cc.betweenRefreshes).Add(jitter),
+		deleteAfter:  old.deleteAfter,
```

Comment:
IIUC this is critical for an entry to get removed, could you please comment why it's here so nobody 'fixes' it?
I know the function comment says it's done on purpose but adding 'WHY' would help.

Reply:
Done.